third_party/blink/renderer/modules/cobalt/h5vcc_runtime/h_5_vcc_runtime.cc

```diff
@@ -67,46 +58,50 @@ void H5vccRuntime::EnsureReceiverIsBound() {
       remote_h5vcc_runtime_.BindNewPipeAndPassReceiver(task_runner));
 }
 
-void H5vccRuntime::RemoteAddListener() {
-  EnsureReceiverIsBound();
-
-  // Bind the receiver for the RemoteListener, this is where the NotifyDeepLink
-  // event is called.
-  auto task_runner =
-      GetExecutionContext()->GetTaskRunner(TaskType::kMiscPlatformAPI);
-  remote_h5vcc_runtime_->AddListener(
-      receiver_.BindNewPipeAndPassRemote(task_runner));
-}
-
 void H5vccRuntime::NotifyDeepLink(const WTF::String& deeplink) {
   MaybeFireDeepLinkEvent(deeplink);
 }
 
-bool H5vccRuntime::AddEventListenerInternal(
+void H5vccRuntime::AddedEventListener(
     const AtomicString& event_type,
-    EventListener* listener,
-    const AddEventListenerOptionsResolved* options) {
-  bool success =
-      EventTarget::AddEventListenerInternal(event_type, listener, options);
-  if (success) {
-    RemoteAddListener();
-  }
-  return success;
+    RegisteredEventListener& registered_listener) {
+  EventTarget::AddedEventListener(event_type, registered_listener);
+  MaybeRegisterMojoListener();
 }
 
 void H5vccRuntime::RemovedEventListener(
     const AtomicString& event_type,
     const RegisteredEventListener& registered_listener) {
   EventTarget::RemovedEventListener(event_type, registered_listener);
-  receiver_.reset();
+  MaybeUnregisterMojoListener();
 }
 
 void H5vccRuntime::Trace(Visitor* visitor) const {
   ScriptWrappable::Trace(visitor);
   ExecutionContextLifecycleObserver::Trace(visitor);
   EventTargetWithInlineData::Trace(visitor);
   visitor->Trace(remote_h5vcc_runtime_);
-  visitor->Trace(receiver_);
+  visitor->Trace(listener_receiver_);
+}
+
+void H5vccRuntime::MaybeRegisterMojoListener() {
+  DCHECK(HasEventListeners(event_type_names::kDeeplink));
+  if (!listener_receiver_.is_bound()) {
+    EnsureRemoteIsBound();
+    // Bind the receiver for the RemoteListener, this is where the
+    // NotifyDeepLink event is called.
+    auto task_runner =
+        GetExecutionContext()->GetTaskRunner(TaskType::kMiscPlatformAPI);
+    remote_h5vcc_runtime_->AddListener(
+        listener_receiver_.BindNewPipeAndPassRemote(task_runner));
+  }
```

Comment:
nit: Prefer early return, and please add the DCHECK I mentioned.
```
  if (listener_receiver_.is_bound()) {
    DCHECK(HasEventListeners(event_type_names::kDeeplink));
    return;
  }
  // The rest.
```

Reply:
Andrew explained this offline. Discard my comnent above the DCHECK, but consider early return.